### DISCUSSION TITLE
How do I delay code from being ran?

### DISCUSSION THREAD

**Discussion by KutikiPlayz** [FabricMC/fabric-api#2983] [Mod Dev Support] [answered]
I'm making a magic mod, one of the spells in it summons 3 ice shards and flings them forward, I need to delay the 2nd and 3rd ice shards so they don't spawn all simultaneously.

I've tried using the timer and delay in a coroutine, both of which crash the game and summon twice the amount I want. I've used a while loop to add delay but that freezes the game whenever I use the spell.

Is there a proper way of doing this or have I just used the timer/coroutines wrong?

Also, I'm coding in kotlin if that matters.

**Comment by warjort** [comment, marked as answer]
You can't do multithreading in minecraft unless you know what you are doing.

You shouldn't be managing the control flow yourself anyway.
In simple terms Minecraft is made up of two threads, a client thread and a server thread.
Things happen when the relevant minecraft thread calls your tick() or other callback method.

The client thread is mainly used for rendering and user interaction.
All other processing should occur on the server thread which is the only place that has the "authoritative state" of the game.

Your "twice the amount" is likely because you are spawning entities on the client AND the server.
Any entity spawned by you on the client will just be a ghost entity - a visual glitch not really part of the game.
You need to spawn entities on the server and let it send updates to the client(s).

I don't know of any vanilla (or modded?) entity that does what you are trying to do.

Normally what are you doing would require the user to hold down the right mouse to cause the item's prolonged usage callbacks to be invoked - i.e. the Item.usageTick(), use(), finishUsing(), onStoppedUsing() methods.
See for example the SpyglassItem, CrossbowItem or one of the other vanilla items where you need to hold down the right button.
